Update rules with required phrases automatically

AyanSinhaMahapatra opened the pull request:
This is a continuation of https://github.com/aboutcode-org/scancode-toolkit/pull/3254 with added required phrases in license rules after review and further manual curations. Also contains improvements in required phrase collection and marking.

Reference: #2637 #2878 

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [ ] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

AyanSinhaMahapatra committed:
Add script for adding required phrases automatically

Add a script which can add required phrases in already existing rules
automatically from required phrases already present in other rules and
license field names. This can be done one license expression at a time.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases automatically for mit rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases at bsd-new rules automatically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases automatically for gpl-2.0

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases for gpl-2.0-only rules automatically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Skip creating required phrase rules for generic licenses

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases for genreic license rules automtaically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases to gpl-1.0-plus rules automatically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases automatically on gpl/lgpl rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add more required phrases automatically in permissive rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases automatically to gpl-3.0 rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases for lgpl rules automatically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases for agpl rules automatically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases to bsd rules automatically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases in lgpl rules automatically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases automatically for cc rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented:
I am pushing shortly a few updates:
- decouple the creation of new rules from updating existing rules in a separate CLI
- ensure we skip more rules in the whole process: any rule that cannot be matched approximately and not only tiny rules, and also false positives
- ensure that no rule get a required phrase addition that would break in the middle of a URL, email, or copyright. This will be done to check that no required phrase injection changes the set of ignorables of a rule and makes the URL not longer a proper URL for instance.
- extend "skipping" the collection of required phrases flag to skip a rule from both required phrases collection AND injection. This allow to handle exceptions more easily.

pombredanne committed:
Improve license detection rules

Do not damage rules with URLs

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve required phrases in paths

Ensure that the leading /usr is not broken with {{ required phrase }}
markers.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve required phrases in paths

Ensure that /usr paths are not broken with {{ required phrase }}
markers.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve required phrases in URLs

Ensure that URLs are not broken with {{ required phrase }} markers.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Correct rule syntax

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Move back required-phrase related code

This is code that belongs to required_phrase.py, not to tokenize.py

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use correct syntax for rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Apache Software Foundation is not required phrase

This creates many false positives.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove duplicate imports

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license models

This helps with required phrases handling, addition and generation

* Add new Rule.source attribute to track the "source" of a license rule
  like when adding a new required phrase to a rule
* Add new Rule.is_tiny computed attribute to ytrack tiny, very small
  rules
* Add new Rule.is_approx_matchable property for rules that can only be
  matched exactly
* Add new Rule.is_generic for rules that contain "generic" licenses
* Support required_phrases-related fields in Rule.validate()
* Update index.py accordingly

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Apply minor cosmetic refactoring

* extract LicenseCache.dump() method
* format code

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Move tests to test_required_phrase.py

These were not tokenizeation-related

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add misc comments to Token class variants

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine is_required_phrase license rules

* Refine is_license_* flags
* Remove {{ tags }}
* Remove or improve ambiguous rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Ensure license index is buildable with extras

Somehow building the index for all languages was failing.
This is fixing this issue.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refactor required phrase spans collection

* This is best moved into tokenize and is also used in models.
* The code is simpler as it does not collect the texts, only the spans
  which is what is used for indexing, matching and required phrases
  tagging.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine is_required_phrase AGPL rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine is_required_phrase GPL3 rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine is_required_phrase GPL rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine is_required_phrase LGPL rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine is_required_phrase Apache rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine is_required_phrase BSD rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add new and improved required phrase rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine rules skipped for required phrase

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Rename Rule.skip_collecting_required_phrases

Renamed to skip_for_required_phrase_generation.
This skips a tagged rule for is_required_phrase rule gen and for
required_phrase rule text update with is_required_phrase rules.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update required phrase generation

* This update decouples the creation of is_required_phrase rules from
  updating existing rules in a separate CLI. This makes it easier to
  control which rule are used as required phrases.

* This now skip to process more rules when adding required phrases to
  existing rules: any rule that cannot be matched approximately is
  skipped and only tiny rules, but also many other rules.

* This checks that no rule get a required phrase added that would
  break in the middle of a URL, email, or copyright. This is done by
  checking that no required phrase injection changes the set of
  ignorables of a rule and could break a URL making it no longer a
  proper URL. Same for emails or copyrights.

* This extends "skipping" the collection of required phrases to skip
  a rule from both required phrases collection for generationg new rules
  AND injection of new required phrases in rule text. This allow to
  handle exceptions more easily.

* The "is_required_phrase" rules creation now creates rules using
  improved content: the case and punctuation of the phrase text are
  preserved; the rule is created as "is_license_reference" which is
  going to be correct in the vast majority of the cases.

* When matched, the "is_required_phrase" rules are treated the same
  as continuous rules and can only be matched exactly.

* The "is_required_phrase" rules are now validated extensively to
  ensure that there is no conflict with other rule flags.

* The code to "trace" the source of a required_phase inject now uses
  the new standard "source" rule field, and the code related to handling
  this field has been simplified.

* Required phrases injection has not yet been tested as working.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Make rules valid

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve required phrase in rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update test results

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Correct rule ignorables

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Set relevance for is_required_phrase rules

This should be always 100% relevant

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Fix duplicate license rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Adjust rules from required phrase tests

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve Apache and commons clause rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve imagemagick rules with required phrases

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve rule selectivity

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Do not add required phrase to license text

Long rules that are a license text become a problem with required
phrases: they no longer can be matched accurately.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license test results

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Consider license clues in gibberrish filter

filter_invalid_matches_to_single_word_gibberish() also considers
license_clues as eligible for gibberish filtering.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use correct dry_run argument name in tests

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update license test expectation

This is a more correct result

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Consider clues in false positive filter

is_candidate_false_positive() now also considers
license_clues as eligible for false poistive filtering.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Fix comment content

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add required phrase to MIT rule

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license test results

Adjust with latest license code detection for required phrases

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license rules and license tests

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Rename new rules to resolve merge conflicts

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Merge latest develop

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve rule ignorables and requried phrases

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Adapt license test expectations

This after merging the latest develop and taking into account updates
with required phrases.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Do not debug trace by default

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license detection results

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Refine license detection rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license detection results

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add new license rule for LGPL

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve GPL detection

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license detection tests

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add new GPL rules

Seen in Debian copyright files

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve MPL tri-license detection

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add new DCO licenses

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add new license detection rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license detection tests

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Relax required phrase filtering

Only process stopwords this for "is_continuous" rules

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Correct license test expectation

Some rules now have a "is_required_phrase" flag

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Correct license rule expression

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update expected license test result

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

petergardfjall commented:
Hey! This PR looks promising as I think it has the potential to greatly reduce false positives.
Just to drop an example, having required phrases throughout the rule set would avoid "silly" mistakes like this:

```json
"rule_url": "https://github.com/nexB/scancode-toolkit/tree/develop/src/licensedcode/data/rules/lgpl-2.1_250.RULE",
"matched_text": "// Licensed under the MIT license. \nnamespace Microsoft.OpenApi.Readers.V2"
```

And I do see quite a lot of these "ghost" detections.

I was just checking in to ask if there is any hope of seeing this PR merged. It has been open for some time.

pombredanne commented:
@petergardfjall re:
> I was j ust checking in to ask if there is any hope of seeing this PR merged. It has been open for some time.

Yes! I am about to have a chat with @AyanSinhaMahapatra on just that.

pombredanne commented:
The longer story is that there are reconciliations to do wrt. the changes I made and  changes that @AyanSinhaMahapatra has not yet pushed.

AyanSinhaMahapatra committed:
Add required phrases for mpl rules automatically

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add required phrases automatically for more rules

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into update-rules-with-required-phrases-auto

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne committed:
Improve ISC rules and tests

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Update test expectations

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve detection of BSD license

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Filter key phrases for continuity correctly

The test for a key phrase uninterrupted continuity must be done in all
cases, whether a matched rule is "continuous", or a full "required
phrase".

Fix test and add new test.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Avoid heisen failures in CI

Some tests running with a single process and threading are randombly
failiing in the CI that run varisou wrapper node-based envt. that then
spawn lower level tools like our test runners.
This sometimes makes the Python process and threading run in corner
cases that fail. This commits moves these issues out of the way to
avoid failing when this fails.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Avoid failures in failure reporting

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
@AyanSinhaMahapatra all green... I am merging now :bow:

codewithfourtix commented:
Hi @AyanSinhaMahapatra and @pombredanne,

I'm Ali Zulfiqar (@codewithfourtix), a CS undergrad applying for GSoC 2026. I'm writing a proposal to extend the work in this PR using NLP/AI to cover rules the heuristic-based approach couldn't reach.

From reading through this PR and the commit history, I can see the current script works by cross-referencing required phrases already present in other rules and license field names essentially a token-overlap heuristic. @pombredanne's commits also added important guards like skipping rules with URLs, emails, and copyrights that would be broken by phrase injection, and the `is_tiny` / `is_approx_matchable` / `skip_required_phrases` flags to handle edge cases.

My proposal is to train a BIO-tagged token classifier (StarCoder2-3B or BERT) on the rules that already have required phrases from this PR as the training dataset, so it can predict required phrase spans in rules the overlap heuristic left behind.

A few specific questions before I finalize my proposal:

1. After this PR merged, roughly how many rules still have no required phrases is there a quick way to count them from the .yml files?
2. The `skip_required_phrases` flag was added to handle exceptions should a model-based approach also respect this flag and never predict on flagged rules?
3. @pombredanne's commit `Apache Software Foundation is not required phrase` showed that some phrases cause false positives even when they seem discriminative. Is there a maintained list of such "banned" phrases the model output should be filtered against?

I've also made 4 PRs and a bug report in the repo (#4781–#4785) to get familiar with the codebase. Happy to share my full draft proposal for early feedback.

Thanks for your time!
